source/imaer-shared/src/main/java/nl/overheid/aerius/shared/emissions/FarmlandEmissionsCalculator.java

```diff
@@ -35,38 +37,37 @@ public FarmlandEmissionsCalculator(final FarmlandEmissionFactorSupplier farmland
 
   public Map<Substance, Double> updateEmissions(final FarmlandEmissionSource farmlandSource) {
     final Map<Substance, BigDecimal> summed = new EnumMap<>(Substance.class);
-    for (final FarmlandActivity activity : farmlandSource.getSubSources()) {
-
-      if (activity instanceof FarmlandGrazingActivity) {
-        updateEmissions((FarmlandGrazingActivity) activity, summed);
-      } else {
-        updateEmissions(activity.getEmissions(), summed);
-      }
+    for (final AbstractFarmlandActivity activity : farmlandSource.getSubSources()) {
+      activity.accept(this, summed);
     }
 
     final Map<Substance, Double> result = new EnumMap<>(Substance.class);
     summed.forEach((key, value) -> result.put(key, value.doubleValue()));
     return result;
   }
 
-  private void updateEmissions(final Map<Substance, Double> emissions, final Map<Substance, BigDecimal> summedEmissions) {
-    emissions.forEach(
+  @Override
+  public void visit(final CustomFarmlandActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
+    activity.getEmissions().forEach(
         (key, value) -> summedEmissions.merge(key, BigDecimal.valueOf(value), (v1, v2) -> v1.add(v2)));
   }
 
-  private void updateEmissions(final FarmlandGrazingActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
-    Map<Substance, Double> repositoryEmissions = farmlandEmissionFactorSupplier.getGrazingEmissionFactors(activity.getGrazingCategoryCode());
-    Map<Substance, Double> userEmissions = activity.getEmissions();
-
-    // Prefer emission factors supplied by the farmlandEmissionFactorSupplier
-    Map<Substance, Double> activityEmissionFactors = repositoryEmissions.isEmpty() ? userEmissions : repositoryEmissions;
+  @Override
+  public void visit(final StandardFarmlandActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
+    final Map<Substance, Double> emissionFactors = farmlandEmissionFactorSupplier
+        .getFarmSourceEmissionFactors(activity.getFarmSourceCategoryCode());
 
     if (activity.getFarmEmissionFactorType() == FarmEmissionFactorType.PER_ANIMAL_PER_DAY) {
```

Comment:
The `FarmEmissionFactorType` shouldn't be present on the activity I think, but it should be retrieved from the `farmlandEmissionFactorSupplier` based on both the `activityCode` and the `farmSourceCategoryCode`? It's not something that is defined by the user, but rather by the definition in the database for the supplied activity code (which translates to some sector) and the actual category.

Reply:
Good idea, I think the `farmSourceCategoryCode` will be enough, since the code is unique in table `farm_source_categories` and each row must contain the emissionFactorType.

Reply:
ah yeah, if that's guaranteed to be unique that should work as well, thought there might be a pkey on sector/code combination, but this works as well.